src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -1641,11 +1645,17 @@ private BoundThisReference BindThis(ThisExpressionSyntax node, DiagnosticBag dia
                 hasErrors = IsRefOrOutThisParameterCaptured(node, diagnostics);
             }
 
-            return ThisReference(node, this.ContainingType, hasErrors);
+            var currentType = this.ContainingType;
+            // PROTOTYPE: Get the type of `this` in a better way. Probably want to modify the above if statement?
+            var typeOfThis = currentType?.ExtensionClassType ?? currentType;
+            return ThisReference(node, typeOfThis, hasErrors);
         }
 
-        private BoundThisReference ThisReference(CSharpSyntaxNode node, NamedTypeSymbol thisTypeOpt, bool hasErrors = false, bool wasCompilerGenerated = false)
+        private BoundThisReference ThisReference(CSharpSyntaxNode node, TypeSymbol thisTypeOpt, bool hasErrors = false, bool wasCompilerGenerated = false)
         {
+            // If we have an extension class here, it means we used ContainingType instead of ReceiverType to get the type.
+            // That was previously okay, since they were always equivalent in the past, but extension classes break that.
+            Debug.Assert(!((thisTypeOpt as NamedTypeSymbol)?.IsExtensionClass ?? false));
```

Comment:
Would the expression be slightly simpler as: `Debug.Assert((o as C)?.IsExtensionClass != true)`?